components/esp_modem/examples/modem_tcp_client/components/extra_tcp_transports/include/mbedtls_wrap.hpp

```diff
@@ -0,0 +1,46 @@
+/*
+ * SPDX-FileCopyrightText: 2023 Espressif Systems (Shanghai) CO LTD
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+#pragma once
+
+#include <utility>
+#include "mbedtls/ssl.h"
+#include "mbedtls/entropy.h"
+#include "mbedtls/ctr_drbg.h"
+#include "mbedtls/error.h"
+
+using  const_buf = std::pair<const unsigned char *, std::size_t>;
+using  buf = std::pair<unsigned char *, std::size_t>;
+
+class Tls {
+public:
+    Tls();
+    bool init(bool is_server, bool verify);
```

Comment:
I would still prefer calling `init()` manually, since there are some usecases in mbedTLS, when you need to setup certificates before initializing or update them and call init again.

The other reason is that with the static builder function we'd have to somehow overload it (or template it), as sometimes it's useful to create a smart pointer and sometimes an old style pointer is needed -- e.g. this case, when we need to pass it to `tcp_transport` layer as C-style API.

Reply:
The first case could be solved with overloads to the builder function, or default parameters, but I would need to understand the use cases before being able to give a concrete suggestion.

For the second part I don't agree with you because we shouldn't tie the lifetime of the type with the builder (beyond its destructor, I'm talking about the pointer scenario). If we return a `T` the user could easily call `new T` and get a raw pointer or pass it to the smart pointer. We can even make the factory return a `T*` and have `nullptr` as a failure case, similar to the C API or, better, a `std::optional<T>`.

Reply:
Ad1) I'm probably not aware of all use cases either, just wanted to keep the flexibility as this is the base class wrapper, that must be inherited and used in a specific case.
One typical use case though: If we're a server, we usually need to load more certificates (i.e. calling  `set_ca_cert()` multiple times) one after another, and that should happen after creation, but before `init()`.

Ad2) Yes, we could template the builder, but that would over-complicate the base class. I can see how the current modem factory has grown which sometimes causes trouble to contributors. (already decided to simplify it)

And there's another problem. This is an abstract class, so it's not supposed to be instantiated (we can probably create a base factory that could be used for derived classes, but again)

My suggestion at the moment would be to move `init()` to the `protected` section, so derived classes can create instances either using factories or constructors with single purpose usec ases (e.g. a simple TLS client wouldn't need the trouble with setting certificates...). What do you think?